File: 20. Valid Parentheses.py
# https://leetcode.com/problems/valid-parentheses/

# Given a string s containing just the characters '(', ')', '{', '}', '[' and
# ']', determine if the input string is valid.

# An input string is valid if:

###############################################################################

# make a tracker list and pop the last element if there is a match

class Solution:
    def isValid(self, s: str) -> bool:
        tracker = []
        for char in s:
            if char in ['(', '[', '{']:
                tracker.append(char)
            elif len(tracker) == 0:
                return False
            elif char == ')' and tracker[-1] != '(':
                return False
            elif char == ']' and tracker[-1] != '[':
                return False
            elif char == '}' and tracker[-1] != '{':
                return False
            else:
                tracker.pop()
        return len(tracker) == 0

File: 20. Valid Parentheses.cpp
// https://leetcode.com/problems/valid-parentheses/

// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']',
// determine if the input string is valid.

// An input string is valid if:

////////////////////////////////////////////////////////////////////////////////

// stack
class Solution {
public:
    bool isValid(string s) {
        stack<char> toRight;
        for (char &c : s) {
            if (c == '(' or c == '[' or c == '{') toRight.push(c);
            else if (toRight.empty()) return false;
            else if (c == ')' and toRight.top() != '(') return false;
            else if (c == ']' and toRight.top() != '[') return false;
            else if (c == '}' and toRight.top() != '{') return false;
            else toRight.pop();
        }
        return (toRight.empty()) ? true : false;
    }
};
